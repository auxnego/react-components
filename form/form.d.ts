import {
  PropsWithChildren,
  //types for html
  AllHTMLAttributes,
  ChangeEventHandler,
  MouseEventHandler
} from "react";

//global types
import {
  ClassNameConfig,
  Config
} from "../global.d";


/** LabelWrapper Component **/

export type LabelWrapperMainProps = {
  label?:string,
  id:string
};

export type LabelWrapperProps = PropsWithChildren &
                                Config<
                                  {root?:string, label?:string},
                                  {root?:object, label?:object}, {}> &
                                AllHTMLAttributes<HTMLElement> &
                                LabelWrapperMainProps;


/** Input Component **/

export type InputMainProps = {
  name?:string,
  pre?:any,
  post?:any,
  type?: "text"   |
         "email"  |
         "search" |
         "number"
};

export type InputProps = LabelWrapperMainProps &
                         PropsWithChildren &
                         ClassNameConfig &
                         InputHTMLAttributes<HTMLInputElement> &
                         Config<
                           {root?:string, label?:string, input?:string, union?:string},
                           {root?:object, label?:object, input?:object, union?:object},
                           {}
                         > &
                         InputMainProps;


/** Checkbox Base Component **/

export type CheckboxBaseMainProps = {
  id:string,
  name?:string,
  value:string,
  onClick:MouseEventHandler
};

export type CheckboxBaseProps = ClassNameOptions &
                                DivHTMLAttibutes<HTMLDivElement> &
                                Config<
                                  {root?:string, input?:string, icon?:string },
                                  {root?:object, input?:object, icon?:object},
                                  {}
                                > &
                                InputMainProps;

/** Radio Base Component **/

export type RadioBaseMainProps = {
  id?:string,
  name:string,
  value:string,
  onChange?:ChangeEventHandler<HTMLInputElement>
};

export type RadioBaseProps = ClassNameOptions &
                             Config<
                               {root?:string, input?:string, sign?:string},
                               {root?:object, input?:object, sign?:object},
                               {}
                             > &
                             InputHTMLAttributes<HTMLInputElement> &
                             RadioBaseMainProps;
