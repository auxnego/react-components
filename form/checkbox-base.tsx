import React, { useState, Attributes, MouseEventHandler } from "react";
import clsx from "clsx";

//types
import {
  CheckboxBaseProps
} from "./form.d";

export default function CheckboxBase(props:CheckboxBaseProps){
  //default props
  let {
    className = "",
    classPrefix = "checkbox-base",
    classless = false,
    icon = null,
    id,
    name = id,
    value,
    onClick,
    setClass = {},
    setProps = {},
    ...more
  } = props;

  setClass = {...setClass, root:"", input:"", icon:"" };
  setProps = {...setProps, root:{}, input:{}, icon:{} };

  //states
  let [$checked, setChecked] = useState<boolean>(!!more.checked);

  //controllers
  let __clickHandler:MouseEventHandler = (e) => {
    if(onClick) onClick(e) ;
    setChecked(!$checked);
  };

  //classNames
  let __classNames = {
    root:[!classless && {
      [`${classPrefix}__root`]:true,
      [`${classPrefix}__root--checked`]:$checked
    }, setClass.root, className],

    input:[!classless && {
      [`${classPrefix}__input`]:true
    }, setClass.input!],

    icon:[!classless && {
      [`${classPrefix}__icon`]:true,
      "fa":true,
      "fa-check":true
    }, setClass.icon!]
  };

  //props
  let __props = {
    root:{
      ...setProps.root!,
      className:clsx(__classNames.root)
    } as Attributes,

    input:{
      ...more,
      ...setProps.input!,
      className:clsx(__classNames.input),
      id,
      name,
      onClick:__clickHandler,
      type:"checkbox",
      value
    } as Attributes,

    icon:{
      ...setProps.icon!,
      className:clsx(__classNames.icon)
    } as Attributes

  };

  //views
  let __icon = icon ? icon : <i {...__props.icon} />;

  return (
    <div {...__props.root}>
      <input {...__props.input} />
      { $checked ? __icon : null}
    </div>
  );
}
