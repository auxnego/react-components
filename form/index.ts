import LabelWrapper from "./label-wrapper";
import Input from "./input";
import CheckboxBase from "./checkbox-base";
import RadioBase from "./radio-base";

export {
  CheckboxBase,
  Input,
  LabelWrapper,
  RadioBase
};
