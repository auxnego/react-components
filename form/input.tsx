import React, { useState, Attributes, FocusEventHandler, ChangeEventHandler } from "react";
import clsx from "clsx";

import Wrapper from "./label-wrapper";

//types
import {
  InputProps
} from "./form.d";

export default function Input(props:InputProps){
  //default props
  let {
    children = null,
    classless = false,
    className = "",
    classPrefix = "input",
    id,
    label = "",
    name = id,
    pre = null,
    post = null,
    setClass = {root:"", label:"", input:"", union:""},
    setProps = {root:{}, label:{}, input:{}, union:{}},
    ...more
  } = props;

  //states
  let [$focused, setFocus] = useState<boolean>(false);
  let [$value, setValue] = useState<string>(more.defaultValue || more.value || "");

  //controllers
  let __blurHandler:FocusEventHandler = (e) => {
    setFocus(false);
  };

  let __changeHandler:ChangeEventHandler<HTMLInputElement> = (e) => {
    setValue(e.target.value);
  };

  let __focusHandler:FocusEventHandler = (e) => {
    setFocus(true);
  };

  //classNames
  let __classNames = {

    root:[!classless && {
      [`${classPrefix}__root`]:true,
      [`${classPrefix}__root--focus`]:$focused,
      [`${classPrefix}__root--filled`]:!!$value
    }, className, setClass.root!],

    label:[!classless && {
      [`${classPrefix}__label`]:true
    }, setClass.label!],

    input:[!classless && {
      [`${classPrefix}`]:true
    }, setClass.input!],

    union:[!classless && {
      [`${classPrefix}__union`]:true
    }, setClass.union!],

  };

  //props
  let __props = {
    //label-wrapper
    wrapper:{
      id,
      label,
      setProps:{
        root:setProps.root!,
        label:setProps.label!
      },
      setClass:{
        root:clsx(__classNames.root),
        label:clsx(__classNames.label)
      }
    },
    //div
    union:{
      ...setProps.union!,
      className:clsx(__classNames.union)
    } as Attributes,
    //input
    input:{
      ...more,
      ...setProps.input!,
      className:clsx(__classNames.input),
      id,
      name,
      onFocus:__focusHandler,
      onBlur:__blurHandler,
      onChange:__changeHandler
    } as Attributes
  };

  //views
  return (
    <Wrapper {...__props.wrapper}>
      <div {...__props.union}>
        {pre}
        <input {...__props.input}/>
        {post}
      </div>
      { children }
    </Wrapper>
  );
}
