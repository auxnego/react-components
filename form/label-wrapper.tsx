import React from "react";
import clsx from "clsx";

//types
import {
  LabelWrapperProps
} from "./form.d";

export default function LabelWrapper(props:LabelWrapperProps){
  //default props
  let {
    children = null,
    id,
    label = id,
    setProps = {},
    setClass = {},
    ...more
  } = props;

  setProps = {...setProps, root:{}, label:{} };
  setClass = {...setClass, root:"", label:"" };
  
  //props
  let __props = {
    root:{
      ...more, 
      ...setProps.root!,
      className:clsx(setClass.root!)
    },
    label:{
      ...setProps.label!,
      className:clsx(setClass.label!)
    }
  };

  return (
    <div { ...__props.root }>
      { label ? <label {...__props.label} htmlFor={id} >{label}</label>
              : null
      }

      { children }

    </div>
  );
}
