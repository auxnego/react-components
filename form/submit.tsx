import React, { Attributes } from "react";
import clsx from "clsx";

import { Button } from "../button";
import LoadingIcon from "../icon/loading";

//types
import {
  SubmitProps
} from "./form.d";

export default function Submit(props:SubmitProps){
  //default props
  let {
    children = null,
    classless = false,
    className = "",
    loading,
    loadingLabel = children,
    setClass = {},
    setProps = {},
    ...more
  } = props;

  setClass = {...setClass, root:"", loading:""};
  setProps = {...setProps, root:{}, loading:{}};
  //classNames
  let __classNames = {

    root:[!classless && {
      "button__submit":true
    }, className, setClass.root!],

    loading:[!classless && {
      "button__submit--loading":true
    }, setClass.loading!]

  };

  //props
  let __props = {

    loading:{
      ...setProps.loading!,
      disabled:true,
      className:clsx(__classNames.loading)
    } as Attributes,

    root:{
      ...more,
      ...setProps.root!,
      type:"submit",
      className:clsx(__classNames.root)
    } as Attributes

  };

  //views
  return loading ? <Button {...__props.loading}><LoadingIcon />{loadingLabel}</Button> : 
                   <Button {...__props.root}>{children}</Button>;
}
