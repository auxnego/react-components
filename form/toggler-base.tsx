import React, { useState, Attributes, MouseEventHandler } from "react";
import clsx from "clsx";

//types
import {
  TogglerBaseProps
} from "./form.d";

export default function TogglerBase(props:TogglerBaseProps){
  //default props
  let {
    classless = false,
    className = "",
    name,
    id = name,
    initialStatus = false,
    onClick,
    setClass = {},
    setProps = {},
    size = "",
    statusText = {},
    offValue = "0",
    onValue = "1",
    ...more
  } = props;

  setClass = {...setClass, root:"", label:"", container:"", input:"", on:"", off:"" };
  setProps = {...setProps, root:{}, label:{}, container:{}, input:{} };

  let __withStatusText:boolean = !!Object.keys(statusText).length;

  //states
  let [$status, setStatus] = useState<boolean>(initialStatus);
  
  //controllers
  let __handleClick:MouseEventHandler<HTMLElement> = (e) => {
    if(onClick) onClick(e);
    setStatus(!$status);
  };

  //classNames
  let __classNames = {

    root:[!classless && {
      "toggler-base__root":true,
      "toggler-base__on":$status
    }, className, setClass.root!],

    label:[!classless && {
      "toggler-base__label":true
    }, setClass.label!],

    container:[!classless && {
      "toggler-base__container":true
    }, setClass.container!],

    input:[!classless && {
      "toggler-base__input":true
    }, setClass.input!],

    on:[!classless && {
      "toggler-base__on":true,
      "toggler-base__on--actived":$status
    }, setClass.on!],

    off:[!classless && {
      "toggler-base__off":true,
      "toggler-base__off--actived":!$status
    }, setClass.off!],
    
  };

  //props
  let __props = {
    root:{
      ...more,
      ...setProps.root!,
      className:clsx(__classNames.root),
      onClick:__handleClick
    } as Attributes,

    container:{
      ...setProps.container!,
      className:clsx(__classNames.container)
    } as Attributes,

    label:{
      ...setProps.label!,
      htmlFor:id,
      className:clsx(__classNames.label)
    } as Attributes,

    input:{
      ...setProps.input!,
      className:clsx(__classNames.input)
    } as Attributes,

    on:{
      className:clsx(__classNames.on)
    } as Attributes,

    off:{
      className:clsx(__classNames.off)
    } as Attributes,

    hiddenInput:{
      type:"hidden",
      name,
      id,
      value: $status ? onValue : offValue
    } as Attributes
  };

  //views
  return (
    <div {...__props.root}>
      <div {...__props.container}>
        <div {...__props.input}>
          <input {...__props.hiddenInput} />
          <div {...__props.on} />
          <div {...__props.off} />
        </div>
      </div>
      {__withStatusText ? <label {...__props.label}>{ $status ? statusText.on : statusText.off }</label> : null}
    </div>
  );
}
