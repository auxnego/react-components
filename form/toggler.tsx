import React, { useState, Attributes, MouseEventHandler } from "react";
import clsx from "clsx";

//types
import {
  TogglerProps
} from "./form.d";

export default function Toggler(props:TogglerProps){
  //default props
  let {
    classless = false,
    className = "",
    name,
    id = name,
    initialStatus = false,
    onClick,
    setClass = {root:"", label:"", input:"", on:"", off:""},
    setProps = {root:{}, label:{}, input:{}},
    size = "",
    statusText = {},
    offValue = "0",
    onValue = "1",
    ...more
  } = props;

  let __withStatusText:boolean = !!Object.keys(statusText).length;

  //states
  let [$status, setStatus] = useState<boolean>(initialStatus);
  
  //controllers
  let __handleClick:MouseEventHandler<HTMLElement> = (e) => {
    if(onClick) onClick(e);
    setStatus(!$status);
  };

  //classNames
  let __classNames = {

    root:[!classless && {
      "toggler__root":true,
      "toggler__on":$status
    }, className, setClass.root!],

    label:[!classless && {
      "toggler__label":true
    }, setClass.label!],

    input:[!classless && {
      "toggler__input":true
    }, setClass.input!],

    on:[!classless && {
      "toggler__on":true,
      "toggler__on--actived":$status
    }, setClass.on!],

    off:[!classless && {
      "toggler__off":true,
      "toggler__off--actived":!$status
    }, setClass.off!],
    
  };

  //props
  let __props = {
    root:{
      ...more,
      ...setProps.root,
      className:clsx(__classNames.root),
      onClick:__handleClick
    } as Attributes,

    label:{
      ...setProps.label,
      htmlFor:id,
      className:clsx(__classNames.label)
    } as Attributes,

    input:{
      ...setProps.input,
      className:clsx(__classNames.input)
    } as Attributes,

    on:{
      className:clsx(__classNames.on)
    } as Attributes,

    off:{
      className:clsx(__classNames.off)
    } as Attributes,

    hiddenInput:{
      type:"hidden",
      name,
      id,
      value: $status ? onValue : offValue
    } as Attributes
  };

  //views
  return (
    <div {...__props.root}>
      <div {...__props.input}>
        <input {...__props.hiddenInput} />
        <div {...__props.on} />
        <div {...__props.off} />
      </div>
      {__withStatusText ? <label {...__props.label}>{ $status ? statusText.on : statusText.off }</label> : null}
    </div>
  );
}
