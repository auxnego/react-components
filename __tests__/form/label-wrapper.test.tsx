import React from "react";
import { create } from "react-test-renderer";

//for test
import { LabelWrapper } from "../../form";

it("With id and label", () => {

  expect(create(
    <LabelWrapper id="hello"/>
  ).toJSON()).toMatchSnapshot();

});

it("Without id and label", () => {

  expect(create(
    <LabelWrapper id=""/>
  ).toJSON()).toMatchSnapshot();

});

it("setProps: add id='root' and id='label'", () => {

  expect(create(
    <LabelWrapper  
      id="" 
      setProps={{root:{id:"root"}, label:{id:"label"}}} 
      label="hello"/>
  ).toJSON()).toMatchSnapshot();

});

it("setClass: className=root and className=label", () => {

  expect(create(
    <LabelWrapper id="hello2" setClass={{root:"root", label:"label"}}/>
  ).toJSON()).toMatchSnapshot();

});

