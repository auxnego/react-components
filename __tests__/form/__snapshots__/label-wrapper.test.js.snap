// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`With id and label 1`] = `
<div
  className=""
>
  <label
    className=""
    htmlFor="hello"
  >
    hello
  </label>
</div>
`;

exports[`Without id and label 1`] = `
<div
  className=""
/>
`;

exports[`setClass: className=root and className=label 1`] = `
<div
  className=""
>
  <label
    className=""
    htmlFor="hello2"
  >
    hello2
  </label>
</div>
`;

exports[`setProps: add id='root' and id='label' 1`] = `
<div
  className=""
>
  <label
    className=""
    htmlFor=""
  >
    hello
  </label>
</div>
`;
